File: src/commands/claim.rs
use crate::{Context, Error, utils, upgrades};
use poise::serenity_prelude as serenity;
use rand::Rng;
use serde::Deserialize;

#[derive(Deserialize)]
pub struct Skin {
    pub title: String,
    pub image_url: String,
    pub colour: String,
    pub footer: String,
    pub thumbnail: String,
}

pub struct RewardResult {
    pub amount: i64,
    pub title: String,
    pub image_url: String,
    pub colour: String,
    pub footer: String,
    pub thumbnail: String,
}

pub enum ClaimStatus {
    Cooldown { title: String, gif_url: String, hours_remaining: f64 },
    BrokenStreak { amount: i64, previous_dinonuggies: i64, previous_streak: i64 },
    Success { amount: i64, title: String, image_url: String, colour: String, footer: String, thumbnail: String, previous_dinonuggies: i64, previous_streak: i64 },
}

const BEKI_COOLDOWN_RESPONSES: [(&str, &str); 7] = [
    ("Beki is currently cooking the next batch of dinonuggies please wait", "https://c.tenor.com/i6sOwD66MAEAAAAC/tenor.gif"),
    ("Beki is having a little bit of an issue. Please hold", "https://c.tenor.com/h6XlgMwYBnkAAAAd/tenor.gif"),
    ("Ah shit i forgottt, hang on a momentt-", "https://c.tenor.com/TYW-RNzp6hEAAAAC/tenor.gif"),
    ("uhhh what is beki doing ?", "https://media.tenor.com/RYGLfSXNIRIAAAAi/frieren.gif"),
    ("Beki fucking dies of exhaustion", "https://c.tenor.com/kU_EwdsrkLkAAAAC/tenor.gif"),
    ("Beki Spins", "https://media.tenor.com/WKPXrrxUvEgAAAAi/frieren-kuru-kuru.gif"),
    ("Beki is trying out new hobbies", "https://c.tenor.com/_33fqJ2mxQUAAAAd/tenor.gif"),
];

pub async fn get_base_amount(ctx: &Context<'_>, user_id: &str, streak: i64) -> anyhow::Result<f64> {
    let user = ctx.data().db.get_user(user_id).await?;
    
    let marriage_benefits = ctx.data().db.get_marriage_benefits(user_id).await?;
    let pokemon_count = ctx.data().db.get_unique_pokemon_count(user_id).await?;

    let log2_credits = if user.credits > 1.0 { user.credits.log2() } else { 0.0 };
    let log2_nuggies = if user.dinonuggies > 1.0 { user.dinonuggies.log2() } else { 0.0 };

    let mut base = (5 + streak) as f64;
    base *= upgrades::get_nuggie_flat_multiplier(user.nuggie_flat_multiplier_level);
    base *= 1.0 + (streak as f64 * upgrades::get_nuggie_streak_multiplier(user.nuggie_streak_multiplier_level));
    base *= 1.0 + (log2_credits * upgrades::get_nuggie_credits_multiplier(user.nuggie_credits_multiplier_level));
    base *= 1.0 + (pokemon_count as f64 * upgrades::get_nuggie_poke_multiplier(user.nuggie_pokemon_multiplier_level));
    base *= 1.0 + (log2_nuggies * upgrades::get_nuggie_nuggie_multiplier(user.nuggie_nuggie_multiplier_level));
    base *= marriage_benefits;

    Ok(base)
}

pub async fn get_reward(ctx: &Context<'_>, user_id: &str, streak: i64) -> anyhow::Result<RewardResult> {
    let user = ctx.data().db.get_user(user_id).await?;
    let multiplier = upgrades::get_multiplier_amount(user.multiplier_amount_level);
    let chance = upgrades::get_multiplier_chance(user.multiplier_rarity_level);
    
    let (_rand_val, skin_key) = {
        let mut rng = rand::thread_rng();
        let val: f64 = rng.gen();
        let key = if val < chance.gold {
            "gold"
        } else if val < chance.gold + chance.silver {
            "silver"
        } else if val < chance.gold + chance.silver + chance.bronze {
            "bronze"
        } else {
            "regular"
        };
        (val, key)
    };

    let base_amount = get_base_amount(ctx, user_id, streak).await?;
    let mult_val = match skin_key {
        "gold" => multiplier.gold,
        "silver" => multiplier.silver,
        "bronze" => multiplier.bronze,
        _ => 1.0,
    };
    let amount = (base_amount * mult_val).ceil() as i64;

    let skin_data_str = std::fs::read_to_string("data/config/skin/claim.json")?;
    let skin_data: serde_json::Value = serde_json::from_str(&skin_data_str)?;
    
    let season = ctx.data().db.get_global_config("season").await?.unwrap_or_else(|| "normal".to_string());
    let resolved_season = if skin_data.get(&season).is_some() { &season } else { "normal" };
    
    let skin_obj = &skin_data[resolved_season][skin_key];
    
    let title = skin_obj["title"].as_str().unwrap_or("")
        .replace("{amount}", &utils::format(amount as f64))
        .replace("{multiplier}", &utils::format(mult_val));
    
    let footer = skin_obj["footer"].as_str().unwrap_or("")
        .replace("{gold}", &utils::format(chance.gold * 100.0))
        .replace("{silver}", &utils::format(chance.silver * 100.0))
        .replace("{bronze}", &utils::format(chance.bronze * 100.0))
        .replace("{goldMultiplier}", &utils::format(multiplier.gold))
        .replace("{silverMultiplier}", &utils::format(multiplier.silver))
        .replace("{bronzeMultiplier}", &utils::format(multiplier.bronze));

    Ok(RewardResult {
        amount,
        title,
        image_url: skin_obj["imageUrl"].as_str().unwrap_or("").to_string(),
        colour: skin_obj["colour"].as_str().unwrap_or("#FFFFFF").to_string(),
        footer,
        thumbnail: skin_obj["thumbnail"].as_str().unwrap_or("").to_string(),
    })
}

/// Claim your daily dinonuggies
#[poise::command(slash_command)]
pub async fn claim(ctx: Context<'_>) -> Result<(), Error> {
    ctx.defer().await?;
    let user_id = ctx.author().id.to_string();
    let user = ctx.data().db.get_user(&user_id).await?;
    let now = chrono::Utc::now().naive_utc();
    
    let cooldown_hours = upgrades::get_beki_cooldown(user.beki_level);
    
    if let Some(last_claimed) = user.dinonuggies_last_claimed {
        let diff = now.signed_duration_since(last_claimed);
        if diff.num_seconds() < (cooldown_hours * 3600.0) as i64 {
            let (t, g) = {
                let mut rng = rand::thread_rng();
                BEKI_COOLDOWN_RESPONSES[rng.gen_range(0..BEKI_COOLDOWN_RESPONSES.len())]
            };
            let hours_rem = cooldown_hours - (diff.num_seconds() as f64 / 3600.0);
            
            ctx.send(poise::CreateReply::default()
                .embed(serenity::CreateEmbed::new()
                    .title(t)
                    .thumbnail("https://drive.google.com/thumbnail?id=1oVDRweQoYLU6YfB01LWZpTFQiBS1fRRa")
                    .description(format!("You can claim your next nuggie in {:.1} hours.", hours_rem))
                    .color(0xFF0000)
                    .image(g)
                    .author(serenity::CreateEmbedAuthor::new("dinonuggie").icon_url("https://drive.google.com/thumbnail?id=1oVDRweQoYLU6YfB01LWZpTFQiBS1fRRa"))
                    .footer(serenity::CreateEmbedFooter::new("dinonuggie").icon_url("https://drive.google.com/thumbnail?id=1oVDRweQoYLU6YfB01LWZpTFQiBS1fRRa")))
            ).await?;
            return Ok(());
        }
        
        if diff.num_seconds() > (2 * 24 * 3600) {
            let base = get_base_amount(&ctx, &user_id, 0).await?;
            let amount = base.ceil() as i64;
            
            ctx.data().db.claim_nuggies(&user_id, amount as f64, true).await?;
            
            ctx.send(poise::CreateReply::default()
                .embed(serenity::CreateEmbed::new()
                    .thumbnail("https://drive.google.com/thumbnail?id=1oVDRweQoYLU6YfB01LWZpTFQiBS1fRRa")
                    .title(format!("{} dinonuggies claimed!", utils::format(amount as f64)))
                    .description(format!("You now have {} dinonuggies. You broke your streak of {} days.", 
                        utils::format(user.dinonuggies + amount as f64),
                        user.dinonuggies_claim_streak
                    ))
                    .color(0x83F28F)
                    .image("https://drive.google.com/thumbnail?id=1oVDRweQoYLU6YfB01LWZpTFQiBS1fRRa")
                    .author(serenity::CreateEmbedAuthor::new("dinonuggie").icon_url("https://drive.google.com/thumbnail?id=1oVDRweQoYLU6YfB01LWZpTFQiBS1fRRa"))
                    .footer(serenity::CreateEmbedFooter::new("dinonuggie").icon_url("https://drive.google.com/thumbnail?id=1oVDRweQoYLU6YfB01LWZpTFQiBS1fRRa")))
            ).await?;
            return Ok(());
        }
    }

    let reward = get_reward(&ctx, &user_id, user.dinonuggies_claim_streak).await?;
    ctx.data().db.claim_nuggies(&user_id, reward.amount as f64, false).await?;

    ctx.send(poise::CreateReply::default()
        .embed(serenity::CreateEmbed::new()
            .thumbnail(&reward.thumbnail)
            .color(utils::parse_hex(&reward.colour))
            .author(serenity::CreateEmbedAuthor::new("dinonuggie").icon_url("https://drive.google.com/thumbnail?id=1oVDRweQoYLU6YfB01LWZpTFQiBS1fRRa"))
            .title(&reward.title)
            .description(format!("You now have {} dinonuggies. You are on a streak of {} days.", 
                utils::format(user.dinonuggies + reward.amount as f64),
                user.dinonuggies_claim_streak + 1
            ))
            .image(&reward.image_url)
            .footer(serenity::CreateEmbedFooter::new(format!("dinonuggie | {}", reward.footer)).icon_url("https://drive.google.com/thumbnail?id=1oVDRweQoYLU6YfB01LWZpTFQiBS1fRRa")))
    ).await?;

    Ok(())
}

File: utils/claim.ts
import { EmbedBuilder, type ChatInputCommandInteraction, type ColorResolvable } from 'discord.js';

import { log } from './log';
import claimSkinConfig from '../data/config/skin/claim.json';
import {
  getNuggieFlatMultiplier,
  getNuggieStreakMultiplier,
  getNuggieCreditsMultiplier,
  getNuggiePokeMultiplier,
  getNuggieNuggieMultiplier,
} from './ascensionupgrades';
import {
  getMultiplierAmount,
  getMultiplierChance,
  getBekiCooldown,
} from './upgrades';
import { format } from './math';
import { withUserLock, userLocks } from './userLock';

const DAY_LENGTH = 24 * 60 * 60 * 1000;
const HOUR_LENGTH = 60 * 60 * 1000;

const BEKI_COOLDOWN_RESPONSES: { title: string; gifUrl: string }[] = [
  { title: 'Beki is currently cooking the next batch of dinonuggies please wait', gifUrl: 'https://c.tenor.com/i6sOwD66MAEAAAAC/tenor.gif' },
  { title: 'Beki is having a little bit of an issue. Please hold', gifUrl: 'https://c.tenor.com/h6XlgMwYBnkAAAAd/tenor.gif' },
  { title: 'Ah shit i forgottt, hang on a momentt-', gifUrl: 'https://c.tenor.com/TYW-RNzp6hEAAAAC/tenor.gif' },
  { title: 'uhhh what is beki doing ?', gifUrl: 'https://media.tenor.com/RYGLfSXNIRIAAAAi/frieren.gif' },
  { title: 'Beki fucking dies of exhaustion', gifUrl: 'https://c.tenor.com/kU_EwdsrkLkAAAAC/tenor.gif' },
  { title: 'Beki Spins', gifUrl: 'https://media.tenor.com/WKPXrrxUvEgAAAAi/frieren-kuru-kuru.gif' },
  { title: 'Beki is trying out new hobbies', gifUrl: 'https://c.tenor.com/_33fqJ2mxQUAAAAd/tenor.gif' },
];

interface MultiplierAmount {
  gold: number;
  silver: number;
  bronze: number;
}

interface MultiplierChance {
  gold: number;
  silver: number;
  bronze: number;
}

interface RewardResult {
  amount: number;
  title: string;
  imageUrl: string;
  webImageUrl?: string;
  colour: string;
  footer: string;
  thumbnail: string;
}

async function getBaseAmount(client: any, uid: string, streak: number): Promise<number> {
  const nuggieFlatMultiplierLevel = await client.db.user.getUserAttr(uid, 'nuggieFlatMultiplierLevel');
  const nuggieStreakMultiplierLevel = await client.db.user.getUserAttr(uid, 'nuggieStreakMultiplierLevel');
  const marriageBenefitsMultiplier = await client.db.marriage.getMarriageBenefits(uid);

  const nuggieCreditsMultiplierLevel = await client.db.user.getUserAttr(uid, 'nuggieCreditsMultiplierLevel');
  const credits = await client.db.user.getUserAttr(uid, 'credits');
  const log2Credits = credits > 1 ? Math.log2(credits) : 0;

  const nuggiePokemonMultiplierLevel = await client.db.user.getUserAttr(uid, 'nuggiePokemonMultiplierLevel');
  const pokemonCount = await client.db.pokemon.getUniquePokemonCount(uid);

  const nuggieNuggieMultiplierLevel = await client.db.user.getUserAttr(uid, 'nuggieNuggieMultiplierLevel');
  const nuggies = await client.db.user.getUserAttr(uid, 'dinonuggies');
  const log2Nuggies = nuggies > 1 ? Math.log2(nuggies) : 0;

  let baseAmount = (5 + streak);
  log(`Base amount: ${baseAmount}`);
  baseAmount *= getNuggieFlatMultiplier(nuggieFlatMultiplierLevel);
  log(`Base amount after flat multiplier: ${baseAmount}`);
  baseAmount *= (1 + streak * getNuggieStreakMultiplier(nuggieStreakMultiplierLevel));
  log(`Base amount after streak multiplier: ${baseAmount}`);
  baseAmount *= (1 + log2Credits * getNuggieCreditsMultiplier(nuggieCreditsMultiplierLevel));
  log(`Base amount after credits multiplier: ${baseAmount}`);
  baseAmount *= (1 + pokemonCount * getNuggiePokeMultiplier(nuggiePokemonMultiplierLevel));
  log(`Base amount after pokemon multiplier: ${baseAmount}`);
  baseAmount *= (1 + log2Nuggies * getNuggieNuggieMultiplier(nuggieNuggieMultiplierLevel));
  log(`Base amount after nuggie multiplier: ${baseAmount}`);
  baseAmount *= marriageBenefitsMultiplier;
  log(`Base amount after marriage benefits: ${baseAmount}`);
  return baseAmount;
}

async function formatReward(
  client: any,
  skinKey: string,
  amount: number,
  multiplier: MultiplierAmount,
  percentage: MultiplierChance,
): Promise<RewardResult> {
  const season = await client.db.globalConfig.getGlobalConfig('season') || 'normal';
  const json = claimSkinConfig as any;
  const resolvedSeason = json[season] ? season : 'normal';
  log(`season: ${season}`);
  log(`resolvedSeason: ${resolvedSeason}`);
  log(`skinKey: ${skinKey}`);
  log(`json[resolvedSeason]: ${JSON.stringify(json[resolvedSeason])}`);
  const skin = json[resolvedSeason][skinKey];
  const goldPercentage = format(percentage.gold * 100, true);
  const silverPercentage = format(percentage.silver * 100, true);
  const bronzePercentage = format(percentage.bronze * 100, true);
  const goldMultiplier = format(multiplier.gold, true);
  const silverMultiplier = format(multiplier.silver, true);
  const bronzeMultiplier = format(multiplier.bronze, true);

  // Determine appropriate multiplier based on skinKey, defaulting to 1.0 for "regular"
  const currentMultiplier = (multiplier as any)[skinKey] !== undefined ? (multiplier as any)[skinKey] : 1.0;

  // Format footer with unique multipliers for each rarity
  const formattedFooter = skin.footer
    .replace(/{gold}/g, goldPercentage)
    .replace(/{silver}/g, silverPercentage)
    .replace(/{bronze}/g, bronzePercentage)
    .replace(/{goldMultiplier}/g, goldMultiplier)
    .replace(/{silverMultiplier}/g, silverMultiplier)
    .replace(/{bronzeMultiplier}/g, bronzeMultiplier);

  return {
    amount,
    title: skin.title.replace('{amount}', format(amount)).replace('{multiplier}', format(currentMultiplier, true)),
    imageUrl: skin.imageUrl,
    webImageUrl: skin.webImageUrl,
    colour: skin.colour,
    footer: formattedFooter,
    thumbnail: skin.thumbnail,
  };
}

async function getAmount(client: any, uid: string, streak: number): Promise<RewardResult> {
  const rand = Math.random();
  const multiplierAmountLevel = await client.db.user.getUserAttr(uid, 'multiplierAmountLevel');
  const multiplierRarityLevel = await client.db.user.getUserAttr(uid, 'multiplierRarityLevel');
  const multiplier = getMultiplierAmount(multiplierAmountLevel);
  const percentage = getMultiplierChance(multiplierRarityLevel);
  log('claiming dinonuggies');

  if (rand < percentage.gold) {
    const amount = Math.ceil(await getBaseAmount(client, uid, streak) * multiplier.gold);
    return formatReward(client, 'gold', amount, multiplier, percentage);
  }
  if (rand < percentage.gold + percentage.silver) {
    const amount = Math.ceil(await getBaseAmount(client, uid, streak) * multiplier.silver);
    return formatReward(client, 'silver', amount, multiplier, percentage);
  }
  if (rand < percentage.gold + percentage.silver + percentage.bronze) {
    const amount = Math.ceil(await getBaseAmount(client, uid, streak) * multiplier.bronze);
    return formatReward(client, 'bronze', amount, multiplier, percentage);
  }
  const amount = Math.ceil(await getBaseAmount(client, uid, streak));
  return formatReward(client, 'regular', amount, multiplier, percentage);
}

type ClaimResult =
  | {
    status: 'cooldown';
    title: string;
    gifUrl: string;
    hoursRemaining: number;
    cooldown: number;
  }
  | {
    status: 'broken_streak';
    amount: number;
    previousDinonuggies: number;
    previousStreak: number;
  }
  | {
    status: 'success';
    amount: number;
    title: string;
    imageUrl: string;
    webImageUrl?: string;
    colour: string;
    footer: string;
    thumbnail: string;
    previousDinonuggies: number;
    previousStreak: number;
  };

async function processClaimInner(client: any, uid: string): Promise<ClaimResult> {
  const now = Date.now();
  const lastClaimedInt = await client.db.user.getUserAttr(uid, 'dinonuggiesLastClaimed');
  const lastClaimed = lastClaimedInt || null;
  const diff = lastClaimed ? now - lastClaimed : DAY_LENGTH;

  const streak = await client.db.user.getUserAttr(uid, 'dinonuggiesClaimStreak');
  const dinonuggies = await client.db.user.getUserAttr(uid, 'dinonuggies');
  const bekiLevel = await client.db.user.getUserAttr(uid, 'bekiLevel');
  const cooldown = getBekiCooldown(bekiLevel);

  if (diff < cooldown * HOUR_LENGTH) {
    const selected = BEKI_COOLDOWN_RESPONSES[Math.floor(Math.random() * BEKI_COOLDOWN_RESPONSES.length)];
    return {
      status: 'cooldown',
      title: selected.title,
      gifUrl: selected.gifUrl,
      hoursRemaining: cooldown - diff / HOUR_LENGTH,
      cooldown,
    };
  }

  if (diff > 2 * DAY_LENGTH) {
    const amount = Math.ceil(await getBaseAmount(client, uid, 0));
    await client.db.user.addUserAttr(uid, 'dinonuggies', amount);
    await client.db.user.setUserAttr(uid, 'dinonuggiesLastClaimed', now);
    await client.db.user.setUserAttr(uid, 'dinonuggiesClaimStreak', 1);
    return {
      status: 'broken_streak',
      amount,
      previousDinonuggies: dinonuggies,
      previousStreak: streak,
    };
  }

  const reward = await getAmount(client, uid, streak);
  await client.db.user.addUserAttr(uid, 'dinonuggies', reward.amount);
  await client.db.user.setUserAttr(uid, 'dinonuggiesLastClaimed', now);
  await client.db.user.addUserAttr(uid, 'dinonuggiesClaimStreak', 1);
  return {
    status: 'success',
    amount: reward.amount,
    title: reward.title,
    imageUrl: reward.imageUrl,
    webImageUrl: reward.webImageUrl,
    colour: reward.colour,
    footer: reward.footer,
    thumbnail: reward.thumbnail,
    previousDinonuggies: dinonuggies,
    previousStreak: streak,
  };
}

function processClaim(client: any, uid: string): Promise<ClaimResult> {
  return withUserLock(userLocks, uid, () => processClaimInner(client, uid));
}

// eslint-disable-next-line max-len
async function handleSuccessfulClaim(client: any, interaction: ChatInputCommandInteraction, newMessage = false): Promise<void> {
  const streak = await client.db.user.getUserAttr(interaction.user.id, 'dinonuggiesClaimStreak');
  const dinonuggies = await client.db.user.getUserAttr(interaction.user.id, 'dinonuggies');
  const now = Date.now();
  const {
    amount, title, imageUrl, colour, footer, thumbnail,
  } = await getAmount(client, interaction.user.id, streak);
  const embed = new EmbedBuilder()
    .setThumbnail(thumbnail)
    .setColor(colour as ColorResolvable)
    .setAuthor({ name: 'dinonuggie', iconURL: 'https://drive.google.com/thumbnail?id=1oVDRweQoYLU6YfB01LWZpTFQiBS1fRRa' })
    .setTitle(title)
    .setDescription(`You now have ${format(dinonuggies + amount)} dinonuggies. You are on a streak of ${streak + 1} days.`)
    .setImage(imageUrl)
    .setFooter({ text: `dinonuggie | ${footer}`, iconURL: 'https://drive.google.com/thumbnail?id=1oVDRweQoYLU6YfB01LWZpTFQiBS1fRRa' });

  await client.db.user.addUserAttr(interaction.user.id, 'dinonuggies', amount);
  await client.db.user.setUserAttr(interaction.user.id, 'dinonuggiesLastClaimed', now);
  await client.db.user.addUserAttr(interaction.user.id, 'dinonuggiesClaimStreak', 1);
  if (newMessage) {
    await interaction.followUp({ embeds: [embed] });
  } else {
    await interaction.editReply({ embeds: [embed] });
  }
}

export {
  getBaseAmount,
  getAmount,
  formatReward,
  handleSuccessfulClaim,
  processClaim,
  BEKI_COOLDOWN_RESPONSES,
  HOUR_LENGTH,
  DAY_LENGTH,
};
export type { ClaimResult };
